test/token/ERC6909/ERC6909.behavior.js

```diff
@@ -0,0 +1,199 @@
+        await this.token.$_mint(this.bruce, secondTokenId, secondTokenAmount);
+      });
+
+      it('transfers to the zero address are blocked', async function () {
+        await expect(
+          this.token.connect(this.alice).transfer(ethers.ZeroAddress, firstTokenId, firstTokenAmount),
+        ).to.be.revertedWithCustomError(this.token, 'ERC6909InvalidReceiver');
+        await expect(this.token.balanceOf(this.alice, firstTokenId)).to.eventually.equal(firstTokenAmount);
+      });
+
+      it('reverts when insufficient balance', async function () {
+        await expect(this.token.connect(this.alice).transfer(this.bruce, firstTokenId, firstTokenAmount + 1n))
+          .to.be.revertedWithCustomError(this.token, 'ERC6909InsufficientBalance')
+          .withArgs(this.alice, firstTokenAmount, firstTokenAmount + 1n, firstTokenId);
+      });
+
+      it('emits event and transfers tokens', async function () {
+        await expect(this.token.connect(this.alice).transfer(this.bruce, firstTokenId, firstTokenAmount))
+          .to.emit(this.token, 'Transfer')
+          .withArgs(this.alice, this.alice, this.bruce, firstTokenId, firstTokenAmount);
+        await expect(this.token.balanceOf(this.alice, firstTokenId)).to.eventually.equal(0);
+        return expect(this.token.balanceOf(this.bruce, firstTokenId)).to.eventually.equal(firstTokenAmount);
+      });
+    });
+
+    describe('transferFrom', function () {
+      beforeEach(async function () {
+        await this.token.$_mint(this.alice, firstTokenId, firstTokenAmount);
+        await this.token.$_mint(this.bruce, secondTokenId, secondTokenAmount);
+      });
+
+      it('transfer from self', async function () {
+        await this.token.connect(this.alice).transferFrom(this.alice, this.bruce, firstTokenId, firstTokenAmount);
+        await expect(this.token.balanceOf(this.alice, firstTokenId)).to.eventually.equal(0);
+        await expect(this.token.balanceOf(this.bruce, firstTokenId)).to.eventually.equal(firstTokenAmount);
+      });
+
+      describe('with approval', async function () {
+        beforeEach(async function () {
+          await this.token.connect(this.alice).approve(this.operator, firstTokenId, firstTokenAmount - 1n);
+          this.tx = await this.token
+            .connect(this.operator)
+            .transferFrom(this.alice, this.bruce, firstTokenId, firstTokenAmount - 1n);
+        });
+
+        it('reverts when insufficient allowance', async function () {
+          await expect(this.token.connect(this.operator).transferFrom(this.alice, this.bruce, firstTokenId, 1))
+            .to.be.revertedWithCustomError(this.token, 'ERC6909InsufficientAllowance')
+            .withArgs(this.operator, 0, 1, firstTokenId);
+        });
+
+        it('should emit transfer event', async function () {
+          await expect(this.tx)
+            .to.emit(this.token, 'Transfer')
+            .withArgs(this.operator, this.alice, this.bruce, firstTokenId, firstTokenAmount - 1n);
```

Comment:
I'd add a check that it doesn't emit an `Approval` event.
```suggestion
            .withArgs(this.operator, this.alice, this.bruce, firstTokenId, firstTokenAmount - 1n)
            .to.not.emit(this.token, 'Approval');
```